web/app/datasets/handle_signals.py

```diff
@@ -13,14 +13,105 @@
+
 import os
+import datetime
+import logging
 
 import requests
+from django.conf import settings
+from django.utils import timezone
+
+from datasets.models import MessageLog
+from datasets.external.base import get_handler
+
+# -- setup logging --
+LOG_FORMAT = '%(asctime)-15s - %(name)s - %(message)s'
+logging.basicConfig(format=LOG_FORMAT, level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+# -- Datapunt internal Signalen in Amsterdam API endpoints --
+#    note: should be set by Ansible or docker-compose.yml
+SIGNALS_API_BASE = os.getenv('SIGNALS_API_BASE', 'https://acc.api.data.amsterdam.nl')
+
+
+def _get_session_with_retries():
+    """
+    Get a requests Session that will retry 5 times on a number of HTTP 500 statusses.
+    """
+    session = requests.Session()
+
+    retries = Retry(
+        total=5,
+        backoff_factor=0.1,
+        status_forcelist=[500, 502, 503, 504],
+        raise_on_status=True
+    )
+    adapter = HTTPAdapter(max_retries=retries)
+    session.mount('https://', adapter)
+    session.mount('http://', adapter)
+
+    return session
+
+
+def _batch_signals(signals):
+    """
+    Access the Signalen in Amsterdam API, retrieve signals.
+    """
+    next_page = SIGNALS_API_BASE + '/signal/'
+
+    with _get_session_with_retries() as session:
+        result = session.get(next_page)
+        yield result.json()['results']
+
+        next_page = results.json()['_links']['next']['href']
+        if next_page == None:
+            raise StopIteration
+
+
+def _call_external_apis(signals):
+    """
+    Call external APIs for each of the signals.
+
+    Note signals are expected as dictionaries, not objects.
+    """
+    for s in signals:
+        # Check local database to see whether this signal was alreay sent to
+        # the relevant external API (if so skip, else update local database).
+        try:
+            entry = MessageLog.objects.get(signal_id=s['signal_id'])
+        except MessageLog.DoesNotExist:
+            logger.debug('Creating entry for {}.'.format(s['signal_id']))
+            entry = MessageLog(
+                signal_id=s['signal_id'],
+                t_entered=timezone.now()
+            )
+            entry.save()
+        else:
+            logger.debug('Retrieved entry for {}.'.format(s['signal_id']))
+            if entry.is_sent:
+                continue
+
+        # Send the signal to the correct API.
+        handler = get_handler(s)
+        success, status = handler.handle(s)
 
+        # Save the status to our local database.
+        entry.is_sent = success
+        entry.status = status
+        entry.t_sent = timezone.now()
+        entry.handler_name = handler.name
 
-def _route_signals(signal):
-    pass # use external_api property to route
+        entry.save()
 
 
 def handle_signals():
-    pass
+    """
+    Entry point (called via manage.py), retrieve and handle signals.
+    """
+    for signals in _batch_signals():
+        _call_external_apis(signals)
```

Comment:
_batch_signals might raise an exception, does this fail the handler or is this catched anywhere else. You do not want the signals-export to crash, it should run more or less as a fault tolerant service isn't it?

Reply:
Then if so, add some tests to show the behaviour when the get fails

Reply:
If this is about the StopIteration error, Python takes care of it. Regarding the crashing of this service - the script could crash without taking down the server (by docker-compose exe-cing). 

I would like to discuss this further as this goes to the design of the service (and the underlying assumptions).

Reply:
The question is, does the server have to crash if the other system is (temporarily) not available? I consider this a perfectly normal situation that has to be handled by the software. The service should only crash when the contract is not met, eg data is received in the wrong format. The service should only crash when it is not able to recover by itself.

Reply:
Server is not crashing; it is the periodic import that may crash. 

A crash here (in the import) is caused by the "signals" API if:
- It is offline (no recovering here, just exit, wait for the next import job to be started).
- It produces errors (again no recovering here, just exit with an appropriate log message, possibly one that is picked up by health monitoring or alerting).
- It works but produces data in a different format than expected. This cannot be fixed by the running program (it takes a programmer).

The service will remain online and serve the Django REST Framework page showing which of the signals were already handled.

Reply:
OK